spec/features/account/backup_codes_spec.rb

```diff
@@ -0,0 +1,39 @@
+require 'rails_helper'
+
+feature 'Backup codes' do
+  before do
+    sign_in_and_2fa_user(user)
+  end
+
+  context 'with backup codes' do
+    let(:user) { create(:user, :with_backup_code) }
+
+    it 'shows backup code section' do
+      expect(page).to have_content(t('account.index.backup_codes_exist'))
+    end
+  end
+
+  context 'without backup codes just phone' do
+    let(:user) { create(:user, :signed_up) }
+
+    it 'does not show backup code section' do
+      expect(page).to have_content(t('account.index.backup_codes_no_exist'))
+    end
+  end
+
+  context 'user clicks regenerate backup codes' do
+    let(:user) { create(:user, :with_backup_code, :with_piv_or_cac) }
+
+    it 'user can click regenerate backup codes' do
```

Comment:
I think you could get rid of this context block, and move this test into the `with backup codes` context, right? You may even be able to save some time on running the specs if you combined this test with the `shows backup code section` to make `shows backup code section and allows used to regenerate codes`.

Also, thoughts on testing that backup codes were actually generated? You save one of the user's backup codes, and then test that it had been deleted after clicking the button, e.g.

```
old_backup_code = user.backup_code_configurations.sample
# do the stuff to regenerate
expect(BackupCodeConfiguration.where(id: old_backup_code.id).any?).to be_false
expect(user.reload.backup_code_configurations.count).to eq(10)
```

Reply:
Thanks, I like the idea of combining the `` shows backup code`` section and ``user can click regenerate backup codes`` section.

As far as testing the generated backup codes section, that seems to be already done in spec/features/two_factor_authentication/backup_code_sign_up_spec.rb.  The generate and regenerate link both take the user to the page that is tested by that file.  But if you feel this needs to be tested in backup_codes_spec as well, I can do that.

Reply:
Yeah, they go to the same place now, but they are not guaranteed to do that forever. We also aren't guaranteed that it won't have some sort of bug when called outside the sign up flow. I think we want some coverage here to make sure we are confident this works at the feature level.